README.md

[![CRAN
Version](http://www.r-pkg.org/badges/version/CliquePercolation)](https://cran.r-project.org/package=CliquePercolation)
[![Downloads](https://cranlogs.r-pkg.org/badges/CliquePercolation)](https://cran.r-project.org/package=CliquePercolation)
[![Travis build
status](https://travis-ci.com/LangeJens/CliquePercolation.svg?branch=master)](https://travis-ci.com/LangeJens/CliquePercolation)
[![DOI](https://joss.theoj.org/papers/10.21105/joss.03210/status.svg)](https://doi.org/10.21105/joss.03210)

# CliquePercolation

**CliquePercolation** is an **R** package that entails multiple
functions related to the clique percolation community detection
algorithms for undirected, unweighted networks and for undirected,
weighted networks. As various sciences analyze structural properties of
networks, the **CliquePercolation** package can be useful in many fields
such as physics, biology, psychology, computer science, or sociology.

## Installation

To use **CliquePercolation**, you first need to install **R** (find
download page [here](https://cran.r-project.org/)). Moreover, I
recommend to use **R** via *RStudio*, an environment that facilitates
the use of **R** in multiple ways (find download page
[here](https://rstudio.com/)).

The developmental version of **CliquePercolation** is available on
GitHub. You can download it from GitHub directly with the help of the
**devtools** package in **R**:

``` r
install.packages("devtools")
devtools::install_github("LangeJens/CliquePercolation", build_vignettes = TRUE)
```

You can download the current stable version from CRAN:

``` r
install.packages("CliquePercolation")
```

## Community Guidelines

If you have recommendations for improving **CliquePercolation**, want to
report issues or problems with the package, or need support, you can
either open an
[issue](https://github.com/LangeJens/CliquePercolation/issues) on GitHub
or contact me directly via e-mail (<lange.jens@outlook.com>). If you
consider contributing to **CliquePercolation**, you can also submit a
[pull request](https://github.com/LangeJens/CliquePercolation/pulls).

## Illustrative Examples

The following includes two examples – one for an undirected, weighted
network and one for an undirected, unweighted network – to illustrate
parts of the package with real data. For a more extensive description of
(a) the clique percolation algorithms for undirected, weighted as well
as undirected, unweighted networks and (b) the workflow of the package
with multiple, fictitious examples, see the package vignette, e.g., by
running `vignette("CliquePercolation")` after installing and loading
(`library(CliquePercolation)`) the package. Furthermore, the
documentation of each function covers various other standalone examples
with fictitious and real data that can be used to verify their
functionality. An overview of all functions and links to their
documentations are available by running `?'CliquePercolation'`.

### The Obama network

The package includes a data set with 10 evaluative reactions toward
Barack Obama from 5,914 participants (for details see `?Obama`). This
data set can be used to estimate a Gaussian Graphical Model (i.e., a
regularized partial correlation network), which is an undirected,
weighted network. Subsequently, the **CliquePercolation** package allows
analyzing the community structure of the network.

First, we will load the data, estimate the network with the package
**bootnet**, and plot the network with the package **qgraph**.

``` r
#load data
data(Obama)

#estimate network
net <- bootnet::estimateNetwork(Obama, default = "EBICglasso", missing = "pairwise")

#plot network
graph <- plot(net, layout = "spring")
```

<img src="README_files/figure-gfm/unnamed-chunk-4-1.png" style="display: block; margin: auto;" />

Second, we optimize *k* and *I* values for running the clique
percolation algorithm. We do so by determining the entropy values of
community partitions for a range of *k* and *I* values and then applying
a permutation test to determine which of these combinations produce
community partitions that are more surprising than already expected by
chance.

``` r
#determine entropy for a range of k and I values
threshold <- cpThreshold(graph, method = "weighted",
                         k.range = 3:4,
                         I.range = seq(0.1, 0.5, 0.01),
                         threshold = "entropy")

#use permutation test to determine which k and I are optimal
permute <- cpPermuteEntropy(graph, cpThreshold.object = threshold,
                            ncores = 2, seed = 4186)
permute
#> 
#> Confidence intervals for entropy values of random permutations of original network
#> 
#> --------------------
#> 
#> User-specified Settings
#> 
#> n = 100 
#> interval = 0.95 
#> CFinder = FALSE 
#> ncores = 2 
#> seed = 4186 
#> 
#> 
#> --------------------
#> 
#> Confidence intervals
#> 
#>  k 95% CI lower 95% CI upper
#>  3        1.409        1.519
#>  4        1.233        1.350
#> 
#> 
#> --------------------
#> 
#> Extracted rows from cpThreshold object
#> 
#>  k Intensity Number.of.Communities Number.of.Isolated.Nodes Entropy.Threshold
#>  3      0.14                     3                        0             1.539
#>  3      0.15                     3                        2             1.985
#>  3      0.16                     3                        2             1.985
#>  3      0.17                     3                        2             1.985
#>  3      0.18                     2                        4             1.571
#>  4      0.14                     2                        3             1.539
#>  4      0.15                     2                        4             1.571
```

Third, we run the clique percolation algorithm with the optimized
values. That is, *k = 3* and *I = 0.14* produce the highest entropy with
the smallest number of isolated nodes that is more surprising than
already expected by chance.

``` r
#run clique percolation algorithm with optimal k and I
cpk3I.14 <- cpAlgorithm(graph, k = 3, I = 0.14, method = "weighted")

#print results overview
cpk3I.14
#> 
#> Results of clique percolation community detection algorithm
#> 
#> --------------------
#> 
#> User-specified Settings
#> 
#> method = weighted 
#> k = 3
#> I = 0.14
#> 
#> --------------------
#> 
#> Results
#> 
#> Number of communities: 3 
#> Number of shared nodes: 3 
#> Number of isolated nodes: 0 
#> 
#> --------------------
#> 
#> For details, use summary() (see ?summary.cpAlgorithm).

#show detailed summary of results
summary(cpk3I.14)
#> 
#> --------------------
#> Communities (labels as identifiers of nodes)
#> --------------------
#> 
#> Community 1 : Mor Led Car Kno Int Hns 
#> Community 2 : Hns Ang Afr 
#> Community 3 : Led Car Hop Prd 
#> 
#> 
#> --------------------
#> Shared nodes (labels as identifiers of nodes)
#> --------------------
#> 
#> Led Car Hns 
#> 
#> 
#> --------------------
#> Isolated nodes (labels as identifiers of nodes)
#> --------------------
#> 
#> NULL
```

Finally, we plot the community partition with three communities and
three shared nodes onto the original network.

``` r
#plot community partition on original network plot
coloredGraph <- cpColoredGraph(graph,
                               list.of.communities = cpk3I.14$list.of.communities.labels,
                               layout = "spring", theme = "colorblind")
```

<img src="README_files/figure-gfm/unnamed-chunk-7-1.png" style="display: block; margin: auto;" />

The plot indicates that feelings and beliefs partly split up in
separate, yet overlapping communities.

### The immunoglobulin network

The package also includes an undirected, unweighted network of
interactions of 1,316 amino-acids (for details see `?immuno`), which we
can also analyze with the **CliquePercolation** package.

First, we will load the data and plot the network with the package
**qgraph**.

``` r
#load data
data(immuno)

#plot network
graph <- qgraph::qgraph(immuno, layout = "spring", vsize = 1)
```

<img src="README_files/figure-gfm/unnamed-chunk-8-1.png" style="display: block; margin: auto;" />

Second, we optimize *k* for running the clique percolation algorithm. We
do so by determining the ratio of the largest to the second largest
community and the chi values for a range of *k* values to determine
whether a community partition leads to a ratio larger than 2 (and if
yes, which is the first community partition to exceed this threshold)
and/or to the highest chi value with a low number of isolated nodes.

``` r
#determine k for clique percolation
threshold <- cpThreshold(graph, method = "unweighted",
                         k.range = 4:5,
                         threshold = c("largest.components.ratio", "chi"))

threshold
#>   k Number.of.Communities Number.of.Isolated.Nodes Ratio.Threshold
#> 1 4                    17                        9        1.311404
#> 2 5                   219                      102        1.038462
#>   Chi.Threshold
#> 1   0.234150780
#> 2   0.005933637
```

Third, we run the clique percolation algorithm with the optimized value.
That is, *k = 4* produces the highest chi with a low number of isolated
nodes.

``` r
#run clique percolation algorithm with optimal k
cpk4 <- cpAlgorithm(graph, k = 4, method = "unweighted")

#print results overview
cpk4
#> 
#> Results of clique percolation community detection algorithm
#> 
#> --------------------
#> 
#> User-specified Settings
#> 
#> method = unweighted 
#> k = 4
#> 
#> --------------------
#> 
#> Results
#> 
#> Number of communities: 17 
#> Number of shared nodes: 56 
#> Number of isolated nodes: 9 
#> 
#> --------------------
#> 
#> For details, use summary() (see ?summary.cpAlgorithm).
```

Finally, we plot the community graph, namely a network in which each
node represents a community and edges indicate the number of nodes the
communities share. Larger nodes represent larger communities.

``` r
#plot community graph
communityGraph <- cpCommunityGraph(cpk4$list.of.communities.numbers,
                                   node.size.method = "proportional",
                                   max.node.size = 10,
                                   theme = "colorblind",
                                   layout = "spring",
                                   repulsion = 0.98)
```

<img src="README_files/figure-gfm/unnamed-chunk-11-1.png" style="display: block; margin: auto;" />

The plot indicates that the interactions of amino-acids in
immunoglobulin are structurally organized in three blocks of
interrelated communities. All three blocks are dominated by a few large
communities.


NEWS
Changes in Version 0.4.0
- cpThreshold now also includes the calculation of fuzzy modularity for weighted and signed networks
- added functions FuzzyMod and SignedFuzzyMod
- added citation information
- updated vignette

Changes in Version 0.3.0
- changes to cpColoredGraph()
    - new argument own.colors allows to specify colors for communities or sets
    - new argument avoid.repeated.mixed.colors allows to avoid that mixed communities are assigned the same color
    - fixed a bug that produced an error message when there was only one pure community
- changes to cpCommunitySizeDistribution
    - new argument test.power.law for testing fit of power-law distribution
- cpAlgorithm, cpThreshold, and cpPermuteEntropy allow a symmetric square matrix as input next to a qgraph object
- cpPermuteEntropy allows using parallel computing to speed up calculation
- new error messages in cpCommunityGraph, cpPermuteEntropy, and cpThreshold
- extension of the package vignette
- added two data sets, Obama and immuno
- cpColoredGraph, cpCommunitySizeDistribution, and cpCommunityGraph now return information beyond the plots invisibly
- added print functions for cpAlgorithm and cpPermuteEntropy and summary function for cpAlgorithm
- added package loading message to point out the vignette

Version 0.2.0
- first version submitted to CRAN

R/cpPermuteEntropy.R
#' Confidence Intervals Of Entropy Based On Random Permutations Of Network
#' 
#' Function for determining confidence intervals of entropy values calculated for
#' community partition from clique percolation based on randomly permuted networks
#' of original network.
#' 
#' @param W A qgraph object or a symmetric matrix; see also \link[qgraph]{qgraph}
#' @param cpThreshold.object A cpThreshold object; see also \link{cpThreshold}
#' @param n number of permutations (default is 100)
#' @param interval requested confidence interval (larger than zero and smaller 1;
#'    default is 0.95)
#' @param CFinder logical indicating whether clique percolation for weighted networks
#'    should be performed as in CFinder ; see also \link{cpAlgorithm}
#' @param ncores Numeric.
#'    Number of cores to use in computing results.
#'    Defaults to \code{parallel::detectCores() / 2} or half of your
#'    computer's processing power.
#'    Set to \code{1} to not use parallel computing
#' @param seed Numeric.
#'    Set seed for reproducible results.
#'    Defaults to \code{NULL}
#'    
#' @return
#'   A list object with the following elements:
#'   \describe{
#'   \item{Confidence.Interval}{a data frame with lower and upper bound of
#'      confidence interval for each \code{k}}
#'   \item{Extracted.Rows}{rows extracted from \code{cpThreshold.object} that are
#'      larger than the upper bound of the specified confidence interval for
#'      each \code{k}}
#'   \item{Settings}{user-specified settings}
#' }
#' 
#' @details
#'   The function generates \code{n} random permutations of the network
#'   specified in \code{W}. For each randomly permuted network, it runs \code{cpThreshold}
#'   (see \link{cpThreshold} for more information) with \code{k} and \code{I} values
#'   extracted from the cpThreshold object specified in \code{cpThreshold.object}.
#'   Across permutations, the confidence intervals of the entropy values are determined
#'   for each \code{k} separately.
#' 
#'   The confidence interval of the entropy values is determined separately for each \code{k}.
#'   This is because larger \code{k} have to produce less communities on average,
#'   which will decrease entropy. Comparing confidence intervals of smaller \code{k} to
#'   those of larger \code{k} would therefore be disadvantageous for larger \code{k}.
#' 
#'   In the output, one can check the confidence intervals of each \code{k}. Moreover,
#'   a data frame is produced that takes the cpThreshold object that was specified in
#'   \code{cpThreshold.object} and removes all rows that do not exceed the upper bound of the
#'   confidence interval of the respective \code{k}.
#' 
#' @examples
#' ## Example with fictitious data
#' 
#' # create qgraph object
#' W <- matrix(c(0,1,1,1,0,0,0,0,
#'               0,0,1,1,0,0,0,0,
#'               0,0,0,0,0,0,0,0,
#'               0,0,0,0,1,1,1,0,
#'               0,0,0,0,0,1,1,0,
#'               0,0,0,0,0,0,1,0,
#'               0,0,0,0,0,0,0,1,
#'               0,0,0,0,0,0,0,0), nrow = 8, ncol = 8, byrow = TRUE)
#' W <- Matrix::forceSymmetric(W)
#' W <- qgraph::qgraph(W)
#' 
#' # create cpThreshold object
#' cpThreshold.object <- cpThreshold(W = W, method = "unweighted", k.range = c(3,4),
#'                                   threshold = "entropy")
#' 
#' # run cpPermuteEntropy with 100 permutations and 95% confidence interval
#' \donttest{
#' results <- cpPermuteEntropy(W = W, cpThreshold.object = cpThreshold.object,
#'                             n = 100, interval = 0.95, ncores = 1, seed = 4186)
#' 
#' # check results
#' results
#' }
#' 
#' ## Example with Obama data set (see ?Obama)
#' 
#' # get data
#' data(Obama)
#' 
#' # estimate network
#' net <- qgraph::EBICglasso(qgraph::cor_auto(Obama), n = nrow(Obama))
#' 
#' # create cpThreshold object
#' \donttest{
#' threshold <- cpThreshold(net, method = "weighted",
#'                          k.range = 3:4,
#'                          I.range = seq(0.1, 0.5, 0.01),
#'                          threshold = "entropy")
#' }
#'                           
#' # run cpPermuteEntropy with 50 permutations and 99% confidence interval
#' \donttest{
#' permute <- cpPermuteEntropy(net, cpThreshold.object = threshold,
#'                             interval = 0.99, n = 50, ncores = 1, seed = 4186)
#' 
#' # check results
#' permute
#' }
#' 
#' @author Jens Lange, \email{lange.jens@@outlook.com} 
#' 
#' @export cpPermuteEntropy

cpPermuteEntropy <- function(W, cpThreshold.object, n = 100, interval = 0.95,
                             CFinder = FALSE, ncores, seed = NULL) {
  
  ###check whether W is a qgraph object
  ###if not check whether matrix is symmetric and convert to qgraph object
  if (!isTRUE(methods::is(W, "qgraph"))) {
    
    #error if W is a matrix but not symmetric
    if (isSymmetric(W) == FALSE) {
      stop("If W is a matrix, it must be symmetric.")
    }
    
    #converts matrix to qgraph if input is not a qgraph object
    W <- qgraph::qgraph(W, DoNotPlot = TRUE)
    
    #### OLD ERROR BEGIN ####
    
    #stop("W (network object) must be a qgraph object.")
    
    #### OLD ERROR END ####
    
  }
  
  #error message if n is not larger than zero
  if (n <= 0) {
    stop("n must be larger than zero.")
  }
  
  #error message if interval is not larger than zero and smaller than 1
  if (!(interval < 1 & interval > 0)) {
    stop("interval must be larger than zero and smaller than 1.")
  }
  
  #set default for missing processing cores
  if(missing(ncores)){
    ncores <- ceiling(parallel::detectCores() / 2)
  }
  
  #for n random permutations of graph, the function runs cpThreshold for k.range and I.range
  #random permutations generated by permuting the upper triangle of the weights matrix and forcing it symmetric
  #for each random permutation, the maximum entropy is extracted for each k and saved in data frame
  #information about parameters in original cpThreshold function is extracted from cpThreshold.object
  if (CFinder == FALSE) {
    if ("Intensity"%in%names(cpThreshold.object)) {method <- "weighted"} else {method <- "unweighted"}
  }
  if (CFinder == TRUE) {
    method <- "weighted.CFinder"
  }
  k.range <- sort(unique(cpThreshold.object$k))
  I.range <- sort(unique(cpThreshold.object$Intensity), decreasing = TRUE)
  
  #initialize iterations
  iterations <- as.list(1:n)
  
  #parallel processing
  cl <- parallel::makeCluster(ncores)
  
  #set seed for reproducibility
  if(!is.null(seed)){
    parallel::clusterSetRNGStream(cl = cl, iseed = seed)
  }
  
  #export variables
  parallel::clusterExport(cl = cl,
                          varlist = c("W", "k.range", "I.range",
                                      "method", "cpThreshold"),
                          envir=environment())
  
  #let user know permutations has started
  message("Permutating...\n", appendLF = FALSE)
  
  #Permutate
  perms <- pbapply::pblapply(
    X = iterations, cl = cl,
    FUN = function(X, W, k.range, I.range, method){
      Wmat <- qgraph::getWmat(W)
      Wmat[upper.tri(Wmat)] <- sample(Wmat[upper.tri(Wmat)])
      Wmat <- Matrix::forceSymmetric(Wmat, "U")
      graph <- qgraph::qgraph(Wmat, DoNotPlot = TRUE)
      
      utils::capture.output(
        results_cpThreshold <- cpThreshold(graph, method = method, k.range = k.range,
                                           I.range = I.range, threshold = "entropy")
      ) #capture.output suppresses display of progress bar from cpThreshold
      
      max_ent <- matrix(nrow = length(k.range), ncol = 2)
      colnames(max_ent) <- c("k", "max")
      
      for (j in 1:length(k.range)) {
        max_ent[j,] <- c(k.range[j], max(results_cpThreshold[results_cpThreshold$k == k.range[j], ]$Entropy.Threshold))
      }
      
      return(max_ent)
      
    },
    W = W, k.range = k.range, I.range = I.range, method = method
  )
  
  parallel::stopCluster(cl)
  
  max_ent <- as.data.frame(do.call(rbind, perms))
  
  #### OLD CODE BEGIN ####
  
  #max_ent <- data.frame(k = numeric(),
  #                      max = numeric())
  
  #counter <- 1
  
  #progress_bar_permute <- utils::txtProgressBar(min = 0, max = n,
  #                                              style = 3) #progress bar definition
  #progress_bar_permute_counter <- 0 #counter for progress bar
  
  #for (i in 1:n) {
  #  Wmat <- qgraph::getWmat(W)
  #  Wmat[upper.tri(Wmat)] <- sample(Wmat[upper.tri(Wmat)])
  #  Wmat <- Matrix::forceSymmetric(Wmat, "U")
  #  graph <- qgraph::qgraph(Wmat, DoNotPlot = TRUE)
    
  #  utils::capture.output(
  #    results_cpThreshold <- cpThreshold(graph, method = method, k.range = k.range,
  #                                        I.range = I.range, threshold = "entropy")
  #  ) #capture.output suppresses display of progress bar from cpThreshold
    
  #  for (j in k.range) {
  #    max_ent[counter, "k"] <- j
  #    max_ent[counter, "max"] <- max(results_cpThreshold[results_cpThreshold$k == j, ]$Entropy.Threshold)
  #    counter <- counter + 1
  #  }
    
    #progress bar update
  #  progress_bar_permute_counter <- progress_bar_permute_counter + 1
  #  utils::setTxtProgressBar(progress_bar_permute, progress_bar_permute_counter)
  #}
  
  #close progress bar
  #close(progress_bar_permute)
  
  #### OLD CODE END ####
  
  
  #determine confidence interval based on normal distribution of extracted entropies for each k
  #extract rows of cpThreshold.object with entropy larger than upper confidence for each k
  results <- data.frame(k = numeric(),
                        l = numeric(),
                        u = numeric())
  extracted_rows <- list()
  counter2 <- 1
  for (k in k.range) {
    m <- mean(max_ent[max_ent$k == k, "max"])
    std <- stats::sd(max_ent[max_ent$k == k, "max"])
    lower_ci <- m - stats::qnorm((1 + interval)/2) * std/sqrt(n)
    upper_ci <- m + stats::qnorm((1 + interval)/2) * std/sqrt(n)
    results[counter2, "k"] <- k
    results[counter2, "l"] <- lower_ci
    results[counter2, "u"] <- upper_ci
    
    rows <- cpThreshold.object[cpThreshold.object$k == k & cpThreshold.object$Entropy.Threshold > upper_ci, ]
    extracted_rows[[counter2]] <- rows
    
    counter2 <- counter2 + 1
  }
  names(results) <- c("k",
                      paste(interval*100, "% CI lower", sep = ""),
                      paste(interval*100, "% CI upper", sep = ""))
  extracted_rows <- Reduce(rbind, extracted_rows)
  
  returned_object <- list(Confidence.Interval = results,
                          Extracted.Rows = extracted_rows,
                          Settings = list(n = n,
                                          interval = interval,
                                          CFinder = CFinder,
                                          ncores = ncores,
                                          seed = seed))
  
  class(returned_object) <- "cpPermuteEntropy"
  
  return(returned_object)
}
